### DISCUSSION TITLE
Doubleclick on taskbar to toggle autohide taskbar

### DISCUSSION THREAD

**Discussion by chip33** [valinet/ExplorerPatcher#389] [Ideas]
I've been using this ahk script compiled as an .exe and auto run at startup for a while now to  toggle autohide taskbar.
I saw recently that a similar function is available in 7+ Taskbar Tweaker but have not tried it.

Maybe you might like to include this functionality in your app as another useful feature.
Here is the script in case anyone finds it useful.

```
#NoEnv  ; Recommended for performance and compatibility with future AutoHotkey releases.
; #Warn  ; Enable warnings to assist with detecting common errors.
SendMode Input  ; Recommended for new scripts due to its superior speed and reliability.
SetWorkingDir %A_ScriptDir%  ; Ensures a consistent starting directory.
#NoTrayIcon
#SingleInstance

; Doubleclick on taskbar to toggle autohide taskbar
; https://www.autohotkey.com/boards/viewtopic.php?t=39123

VarSetCapacity(APPBARDATA, A_PtrSize=4 ? 36:48)
~LButton::
   If (A_PriorHotKey = A_ThisHotKey and A_TimeSincePriorHotkey < 500)
   {
      WinGetClass, Class, A
      If ( Class = "Shell_TrayWnd" )

	NumPut(DllCall("Shell32\SHAppBarMessage", "UInt", 4 ; ABM_GETSTATE
											, "Ptr", &APPBARDATA
											, "Int")
	? 2:1, APPBARDATA, A_PtrSize=4 ? 32:40) ; 2 - ABS_ALWAYSONTOP, 1 - ABS_AUTOHIDE
	, DllCall("Shell32\SHAppBarMessage", "UInt", 10 ; ABM_SETSTATE
									   , "Ptr", &APPBARDATA)
   }
return

```

Edit: For reference here is the [MS Documentation](https://docs.microsoft.com/en-us/windows/win32/shell/abm-setstate).

**Comment by chip33** [comment]
Hi,
I just wondered what you feel about this feature. For me the compiled ahk script fulfils a really useful function that MS left out.
I like to read your comments where you explain your thinking so wondered what you think about this one.
I read your comment recently where you were wishing more people would step forward to work on the mess MS have made of this part of Win 11 and agree. I would do more if I had the time but my coding is just nowhere near the level needed.
I do appreciate the staggering amount of work you have done, and just to make clear , I'm fine if this is not included just wondered what you feel about the idea.

**Reply by valinet** [reply]
Well, I have implemented it. I pushed the commit, a pre-release will be generated, you can download that and test it. At some intervals, the staging pre-release versions will become releases, meaning they get pushed to all users via the update mechanism. It's an interesting feature, although yeah, idk, many things are interesting... things like this could be improved endlessly... anyway, I like it, so it's in there now. Thanks.

> I would do more if I had the time but my coding is just nowhere near the level needed.

Yeah, idk, time is always an issue, believe me, you always have to trade off doing some stuff instead of some other, I also get the same amount of hours per day as you do. Also, I encourage you to check the commit, it's WAY easier to write properly in C than struggle yourself with AHK. That's my opinion. For the life in me I will never understand this scare people generally have, where they will use all sorts of tools and scripting languages and whatever just to avoid springing up a compiler and writing a proper native app.

> `A_PtrSize=4 ? 32:40)`

See? Why struggle like that...? Isn't `APPBARDATA abd; abd.cbSize = sizeof(APPBARDATA);` just easier on the eyes, and easier to understand (let's forget for a moment that `APPBARDATA` screams at you, which is not that big of a deal, depends how good your hearing is =)))? Not to mention you do not need to hardcode anything, because you have a compiler and an SDK with headers that help you with that. I mean, not the fact that your original post had bad formatting, I am not talking about that, but in general, AHK has a terrible looking syntax imo. And besides very limited use cases, I never understood the appeal for scripting languages, more so today when most full fledged programming languages come with some runtime support that JIT compiles them for whatever architecture they run on, so no more "but it's cross platform" argument for scripts either. Besides, let's be real, everyone uses x86 anyway. And yeah, I am not living under a rock, I know ARM exists, but for these kind of tools like the script you showed, running them in emulation would be just as fine anyway...

Thanks

**Reply by valinet** [reply]
Also, forgot to mention, thanks for posting the snippet, it's much easier to implement anything when you have something to start your work on, if it's something concrete like here, it's actually even better.

**Reply by valinet** [reply]
![image](https://user-images.githubusercontent.com/6503598/142663868-d4c381ca-cbf5-45a1-94b8-93f2a7201c8f.png)

**Reply by chip33** [reply]
Hi, smiles, I agree,
All the best , and thanks for putting it in. Will look at the commit.
:)

**Comment by chip33** [comment]
When I run ep_setup.exe version 22000.318.38.4 (or 22000.318.38.3) I briefly get the new Properties GUI showing
Version  22000.318.38.4 in the About tab , then shortly after there is a refresh and the GUI is back to  22000.318.38.2 as are all the files in Program Files\ExplorerPatcher and dxgi.dll in Windows directory.
The notification then pops up for the first time saying I've installed  22000.318.38.2
Could well be something my end, I briefly had it registered as shell extension in the past but that was off when I tried to update.
Just letting you know in case it isn't me. I've tried rebooting, uninstalling, etc but it always happens.

**Reply by valinet** [reply]
It's simple why it happens:

1. You disabled UAC.
2. You set it to auto update.

Thus, it downloads the latest **release** and installs that. Staging builds (pre-releases) are ignored. Disable auto updates if you run those, or set to notify only at least.

**Comment by chip33** [comment]
ah yes :) UAC was disabled rather quickly after installing windows 11 .
and I was on prompt to install.
Thanks .

**Reply by chip33** [reply]
PS . the autohide toggle works beautifully of course and the code is neat.

**Reply by doctorwizz** [reply]
Double clicking on the quick launch side of the taskbar does not work to auto hide. Other side works. 
Kinda like when I reported properties did not work on quick launch side. Remember that?

**Reply by plofhaan** [reply]
For me this feature of EP doesn't start to hide the taskbar until after I perform these steps:
1. restore or minimize the currently maximized window 
2. maximize the window again.

Also, if I have notepad open (not maximized, just visible on the desktop) while my Firefox browser window either is minimized to taskbar or is also visible on the dektop, then if I double-click on the taskbar it causes the Firefox window to be maximized.

**Reply by kichrot** [reply]
In this case, these are not EP flaws, but flaws in the API implementation code of WINDOWS 11 itself.
In my program, the functions implemented on the WIN API, to regulate the auto-closing mode of the taskbar, work perfectly on WINDOWS 10 and work extremely poorly on WINDOWS 11.
I’ve been fighting for a month, but I can’t get a normal job at WINDOWS 11 yet.
So that we will have to put up until they correct the implementation of the WIN API .

**Comment by doctorwizz** [comment]
Didn't meant to reply to chip, just to thread.

**Comment by chip33** [comment]
I just honestly have not noticed any difference between its working on windows 11 and on windows 10. 
And as has been suggested we can check if this is the same behaviours as seen when using Windows Taskbar Autohide feature itself or by trying it with 7+ Taskbar Tweaker.
Even when using windows 10 and its own autohide taskbar feature there have always been instances where some windows or applications at least are obscured by the bar initially but it is just a part of how it works and the advantage of being able to chose a constant taskbar or an autohiding one for different work flows and scenarios outweighs that.

**Reply by plofhaan** [reply]
Of course there exist a number of workflows and scenarios where the ability outweighs that. That is not the problem. Just that, for example, when I have _disabled_ the autohide taskbar mode in EP, I would expect not to see negative side effects such as when I start GIMP so that it maximizes its window and then I click to open the File menu in it, the window very briefly animates to minimize before it immediately maximizes again.

**Reply by valinet** [reply]
So what’s the problem here? Can you provide some meaningful report? You messed with that, disabled it and now windows start doing random stuff or what? Does an Explorer restart fix it? Does a reboot fix it?

I will most probably take this out as it’s basically impossible to test and I’d rather have people blame 7+TT if it happens there as well rather than receiving an issue about it every couple of days. Especially considering the code being posted here, I don’t understand what the expectancy for me to do is, idk why this API is so weird; 7+TT uses some undocumented one that seems to do the same thing, I can implement with that instead, but NO ONE bothers to check that and provide a complete report, because yelling here and blaming EP for global warming, the ocean pollution and everything else is always easier than spending a few minutes to look a bit on the code and/or investigate it.

Would taking out this functionality fix it for you, if not able to provide any other info?

**Reply by valinet** [reply]
Do you have reopen saved windows at logon or something like that?

**Reply by valinet** [reply]
There was a similar thread a few days ago. I don’t have the time to test all the scenarios in the world. I don’t use this feature. As @kichrot said above, maybe the API behaves weirdly in Windows 11. Either the undocumented one from 7+TT works better and we switch to that, either it’s broken and I won’t investigate it more than that, so propose what to do, keep it in there for users where it works or just remove it.

If you want to help this project, try to do a bit more than the standard “it does not work here, let valinet investigate it”. I don’t have the time to do all the investigations in the world. As I suggested to the guy in the other thread, try with 7+TT and report back how it behaves with that. He did not want to do it, as he maintains a clean look for the OS and using that for 5 minutes was hurting for its workflow. If one cannot dedicate 5 minutes to helping with this project, then I can’t do much more either, my time is as valuable as everyone else’s, this expectation that “I have a problem, I will throw it on the developer’s table to fix it” leads nowhere, I cannot test, reproduce and fix all the things in the world without more than that help from the party facing the issue. And as I said, if the shell behaves weird in its own when using this in 7+TT as well, there’s not much I can do, I can take this functionality out. If the implementation from 7+TT works well, we switch to that and call it a day, simple.

**Reply by kichrot** [reply]
valinet, I checked the implementation of 7 + TT. All artifacts of the behavior of windows, when switching the mode of autocreting the taskbar remain. In addition, new ones are added, associated with the violation of the location of the shortcuts on the desktop, when switching the mode of autocliding the taskbar.
So, this is clearly a heating code of the WINDOWS 11 itself, not third-party programs.

Please do not remove the function of switching the mode of autochiding the taskbar from the EP, even with the available errors of WINDOWS 11 behavior, it is very convenient and useful.
Thank you for your disinterested work.

**Comment by plofhaan** [comment]
I found out that the problem I had with GIMP is not caused by EP, as the same problem also occurs if I uninstall EP and apply the UndockingDisabled registry hack after that. Rebooting and/or uninstalling TranslucentTB (UWP version) made no difference to this problem. With the registry hack, the problem can be fixed by setting the taskbar to autohide. AFAIK the only other solution (that is, if it can be called that) is to just stick with the Windows 11 taskbar instead. I don't know if this info is useful to anyone, but anyway... I just thought it might be.

EDIT: If the GIMP window position and size are making the window partially offscreen, then the problem occurs, even if the window is maximized after that. Whereas if the window position and size are making it completely visible within the edges of the screen, the problem appears to be gone. Nice! :D

**Comment by valinet** [comment]
I have 2 pieces of news regarding this functionality, one bad and one good:

* The bad: I found out the reason this might fail sometimes: it is because the taskbar "window" does not receive a secondary click (so it does not realize a double click has occurred) when the taskbar is not locked. The reason for this seems to be the way `explorer.exe` detects that the user clicked and holds down the mouse button in order to drag the taskbar to the new position. For people where this setting failed, it might have been because of this. Unfortunately, a fix is not that easy: you cannot predict the future. I cannot know whether a click will be shortly followed by another one (so it is a double click), so that I prevent `explorer` from activating its "taskbar drag" code path and preventing the second click from reaching my hook. If I go through and do not deliver the "mouse up" messages to it, of course I brake taskbar dragging to reposition it. All in all, idk how to solve this right now, it seems to be more trouble than it is worth, so for now, I will label the feature accordingly in the UI, as working only when the taskbar is locked and leave it as is. Maybe some time in the future, I or someone else figures a hack to make this work in the taskbar non-locked scenario.
* The good: I made this tweak work with the stock (Windows 11) primary and secondary taskbar.

I will publish the changes soon. Thanks.

**Reply by chip33** [reply]
The ahk script approach of detecting a double click and seeing if it is on the taskbar "window" posted at top of this thread works reliably for me with EP 22000.376.40.21 with the win 10 taskbar locked or unlocked. 
I can still click on the taskbar and pull it to a new location.
The script can be run and tested and I'm hoping this method can somehow be ported over to resolve the issue. 

Snippet:
```
~LButton::
If (A_PriorHotKey = A_ThisHotKey and A_TimeSincePriorHotkey < 500)
{
 WinGetClass, Class, A
If ( Class = "Shell_TrayWnd" ) ....
}
```

**Reply by valinet** [reply]
That is something else than what I do in my code. Check it out. I don’t use a global mouse hook as AHK does. It’s a difference. Without a non-interfering global hook, it can’t really be done, I told you above why. I don’t feel like setting a global hook just for this, idk, should have thought better before deciding to implement this, but like, I didn’t know all the story before trying it. I moved the option to Advanced section for now, surprisingly, it works much better with the Windows 11 taskbar, but it is subject for removal, using a more flexible solution like AHK is preferable in this case.

**Reply by chip33** [reply]
Yes, I also did not foresee any issue when I suggested it.
Just an idea but can you look at the implementation in 7+Taskbar Tweaker to see if it is done with a global hook,  I just tried 7+TT and it works with Taskbar unlocked.

**Reply by valinet** [reply]
As I said, everything depends on the implementation.

**Reply by Arora-Sir** [reply]
Double-clicking on the taskbar was working fine till now with this script, until the last Windows 11 Insider update! If someone knows some alternative, please let me know!

```
; MouseLButton DoubleClick Show/Hide Taskbar;
~LButton::DoubleClick(hide := !hide) ;{ <-- Double Click Functions

DoubleClick(action)
{
    If (A_PriorHotKey = A_ThisHotKey and A_TimeSincePriorHotkey < 500)
    {
        WinGetClass, Class, A

        ; Show/Hide Taskbar on Tripple click on taskbar
        If Class = Shell_TrayWnd ; or ( Class = "Progman" )
        {
            static ABM_SETSTATE := 0xA, ABS_AUTOHIDE := 0x1, ABS_ALWAYSONTOP := 0x2
            VarSetCapacity(APPBARDATA, size := 2*A_PtrSize + 2*4 + 16 + A_PtrSize, 0)
            NumPut(size, APPBARDATA), NumPut(WinExist("ahk_class Shell_TrayWnd"), APPBARDATA, A_PtrSize)
            NumPut(action ? ABS_AUTOHIDE : ABS_ALWAYSONTOP, APPBARDATA, size - A_PtrSize)
            DllCall("Shell32\SHAppBarMessage", UInt, ABM_SETSTATE, Ptr, &APPBARDATA)
            Return
        }
    }
    return
}
```

**Comment by Arora-Sir** [comment]
Double-clicking on the taskbar was working fine till now with this script, until the last Windows 11 Insider update! If someone knows some alternative, please let me know!

```
; MouseLButton DoubleClick Show/Hide Taskbar;
~LButton::DoubleClick(hide := !hide) ;{ <-- Double Click Functions

DoubleClick(action)
{
    If (A_PriorHotKey = A_ThisHotKey and A_TimeSincePriorHotkey < 500)
    {
        WinGetClass, Class, A

        ; Show/Hide Taskbar on Tripple click on taskbar
        If Class = Shell_TrayWnd ; or ( Class = "Progman" )
        {
            static ABM_SETSTATE := 0xA, ABS_AUTOHIDE := 0x1, ABS_ALWAYSONTOP := 0x2
            VarSetCapacity(APPBARDATA, size := 2*A_PtrSize + 2*4 + 16 + A_PtrSize, 0)
            NumPut(size, APPBARDATA), NumPut(WinExist("ahk_class Shell_TrayWnd"), APPBARDATA, A_PtrSize)
            NumPut(action ? ABS_AUTOHIDE : ABS_ALWAYSONTOP, APPBARDATA, size - A_PtrSize)
            DllCall("Shell32\SHAppBarMessage", UInt, ABM_SETSTATE, Ptr, &APPBARDATA)
            Return
        }
    }
    return
}
```
